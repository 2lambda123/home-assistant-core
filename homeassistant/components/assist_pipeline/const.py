"""Constants for the Assist pipeline integration."""
DOMAIN = "assist_pipeline"

DATA_CONFIG = f"{DOMAIN}.config"

CONF_PIPELINE_TIMEOUT = "pipeline_timeout"
DEFAULT_PIPELINE_TIMEOUT = 60 * 5  # seconds

CONF_WAKE_WORD_TIMEOUT = "wake_word_timeout"
DEFAULT_WAKE_WORD_TIMEOUT = 3  # seconds

CONF_DEBUG_RECORDING_DIR = "debug_recording_dir"
