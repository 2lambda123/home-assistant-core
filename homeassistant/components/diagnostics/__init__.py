"""The Diagnostics integration."""
from __future__ import annotations

from collections import defaultdict
from collections.abc import Callable, Coroutine
from dataclasses import asdict as dataclass_asdict, dataclass
from http import HTTPStatus
import json
import logging
from typing import Any, Protocol

from aiohttp import web
import voluptuous as vol

from homeassistant.components import http, websocket_api
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers import integration_platform
from homeassistant.helpers.device_registry import DeviceEntry, async_get
from homeassistant.helpers.json import ExtendedJSONEncoder
from homeassistant.helpers.system_info import async_get_system_info
from homeassistant.helpers.typing import ConfigType
from homeassistant.loader import async_get_custom_components, async_get_integration
from homeassistant.util.json import (
    find_paths_unserializable_data,
    format_unserializable_data,
)

from .const import DOMAIN, REDACTED, DiagnosticsSubType, DiagnosticsType
from .util import async_redact_data

__all__ = ["REDACTED", "async_redact_data"]

_LOGGER = logging.getLogger(__name__)


@dataclass
class DiagnosticsSubscriptionSupport:
    """Describe subscriptions supported by the platform."""

    config_entry: bool
    domain: bool


@dataclass
class DiagnosticsPlatformData:
    """Diagnostic platform data."""

    config_entry_diagnostics: Callable[
        [HomeAssistant, ConfigEntry], Coroutine[Any, Any, Any]
    ] | None
    device_diagnostics: Callable[
        [HomeAssistant, ConfigEntry, DeviceEntry], Coroutine[Any, Any, Any]
    ] | None
    subscription_support: DiagnosticsSubscriptionSupport


class DiagnosticsData:
    """Diagnostic data."""

    def __init__(self) -> None:
        """Initialize diagnostic data."""
        self.platforms: dict[str, DiagnosticsPlatformData] = {}
        self.domain_subscriptions: defaultdict[
            str, set[tuple[websocket_api.ActiveConnection, int]]
        ] = defaultdict(set)
        self.config_entry_subscriptions: defaultdict[
            str, set[tuple[websocket_api.ActiveConnection, int]]
        ] = defaultdict(set)


async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
    """Set up Diagnostics from a config entry."""
    hass.data[DOMAIN] = DiagnosticsData()

    await integration_platform.async_process_integration_platforms(
        hass, DOMAIN, _register_diagnostics_platform
    )

    websocket_api.async_register_command(hass, handle_info)
    websocket_api.async_register_command(hass, handle_get)
    websocket_api.async_register_command(hass, handle_subscribe_diagnostics)
    hass.http.register_view(DownloadDiagnosticsView)

    return True


class DiagnosticsProtocol(Protocol):
    """Define the format that diagnostics platforms can have."""

    async def async_get_config_entry_diagnostics(
        self, hass: HomeAssistant, config_entry: ConfigEntry
    ) -> Any:
        """Return diagnostics for a config entry."""

    async def async_get_device_diagnostics(
        self, hass: HomeAssistant, config_entry: ConfigEntry, device: DeviceEntry
    ) -> Any:
        """Return diagnostics for a device."""

    @callback
    def async_supports_subscription(self) -> DiagnosticsSubscriptionSupport:
        """Return if the platform supports subscribing to diagnostics."""


async def _register_diagnostics_platform(
    hass: HomeAssistant, integration_domain: str, platform: DiagnosticsProtocol
) -> None:
    """Register a diagnostics platform."""
    diagnostics_data: DiagnosticsData = hass.data[DOMAIN]
    subscription_support: DiagnosticsSubscriptionSupport
    if hasattr(platform, "async_supports_subscription"):
        subscription_support = platform.async_supports_subscription()
    else:
        subscription_support = DiagnosticsSubscriptionSupport(False, False)

    diagnostics_data.platforms[integration_domain] = DiagnosticsPlatformData(
        getattr(platform, "async_get_config_entry_diagnostics", None),
        getattr(platform, "async_get_device_diagnostics", None),
        subscription_support,
    )


@websocket_api.require_admin
@websocket_api.websocket_command({vol.Required("type"): "diagnostics/list"})
@callback
def handle_info(
    hass: HomeAssistant, connection: websocket_api.ActiveConnection, msg: dict
) -> None:
    """List all possible diagnostic handlers."""
    diagnostics_data: DiagnosticsData = hass.data[DOMAIN]
    result = [
        {
            "domain": domain,
            "handlers": {
                DiagnosticsType.CONFIG_ENTRY: info.config_entry_diagnostics is not None,
                DiagnosticsSubType.DEVICE: info.device_diagnostics is not None,
            },
            "supports_subscription": dataclass_asdict(info.subscription_support),
        }
        for domain, info in diagnostics_data.platforms.items()
    ]
    connection.send_result(msg["id"], result)


@websocket_api.require_admin
@websocket_api.websocket_command(
    {
        vol.Required("type"): "diagnostics/get",
        vol.Required("domain"): str,
    }
)
@callback
def handle_get(
    hass: HomeAssistant, connection: websocket_api.ActiveConnection, msg: dict
) -> None:
    """List all diagnostic handlers for a domain."""
    domain = msg["domain"]
    diagnostics_data: DiagnosticsData = hass.data[DOMAIN]

    if (info := diagnostics_data.platforms.get(domain)) is None:
        connection.send_error(
            msg["id"], websocket_api.ERR_NOT_FOUND, "Domain not supported"
        )
        return

    connection.send_result(
        msg["id"],
        {
            "domain": domain,
            "handlers": {
                DiagnosticsType.CONFIG_ENTRY: info.config_entry_diagnostics is not None,
                DiagnosticsSubType.DEVICE: info.device_diagnostics is not None,
            },
            "supports_subscription": dataclass_asdict(info.subscription_support),
        },
    )


@websocket_api.websocket_command(
    {
        vol.Required("type"): "diagnostics/subscribe",
        vol.Required("domain"): str,
        vol.Optional("config_entry"): str,
    }
)
@callback
def handle_subscribe_diagnostics(
    hass: HomeAssistant, connection: websocket_api.ActiveConnection, msg: dict[str, Any]
) -> None:
    """Subscribe to diagnostic messages."""

    diagnostics_data: DiagnosticsData = hass.data[DOMAIN]

    msg_id = msg["id"]
    domain = msg["domain"]
    config_entry = msg.get("config_entry")
    diagnostics_data.domain_subscriptions[domain].add((connection, msg_id))
    if config_entry:
        diagnostics_data.config_entry_subscriptions[config_entry].add(
            (connection, msg_id)
        )

    @callback
    def cancel_subscription() -> None:
        diagnostics_data.domain_subscriptions[domain].remove((connection, msg["id"]))
        if config_entry:
            diagnostics_data.config_entry_subscriptions[config_entry].remove(
                (connection, msg_id)
            )

    connection.subscriptions[msg["id"]] = cancel_subscription

    connection.send_message(websocket_api.result_message(msg["id"]))


async def _async_get_json_file_response(
    hass: HomeAssistant,
    data: Any,
    filename: str,
    domain: str,
    d_id: str,
    sub_id: str | None = None,
) -> web.Response:
    """Return JSON file from dictionary."""
    hass_sys_info = await async_get_system_info(hass)
    hass_sys_info["run_as_root"] = hass_sys_info["user"] == "root"
    del hass_sys_info["user"]

    integration = await async_get_integration(hass, domain)
    custom_components = {}
    all_custom_components = await async_get_custom_components(hass)
    for cc_domain, cc_obj in all_custom_components.items():
        custom_components[cc_domain] = {
            "version": cc_obj.version,
            "requirements": cc_obj.requirements,
        }
    try:
        json_data = json.dumps(
            {
                "home_assistant": hass_sys_info,
                "custom_components": custom_components,
                "integration_manifest": integration.manifest,
                "data": data,
            },
            indent=2,
            cls=ExtendedJSONEncoder,
        )
    except TypeError:
        _LOGGER.error(
            "Failed to serialize to JSON: %s/%s%s. Bad data at %s",
            DiagnosticsType.CONFIG_ENTRY.value,
            d_id,
            f"/{DiagnosticsSubType.DEVICE.value}/{sub_id}"
            if sub_id is not None
            else "",
            format_unserializable_data(find_paths_unserializable_data(data)),
        )
        return web.Response(status=HTTPStatus.INTERNAL_SERVER_ERROR)

    return web.Response(
        body=json_data,
        content_type="application/json",
        headers={"Content-Disposition": f'attachment; filename="{filename}.json.txt"'},
    )


class DownloadDiagnosticsView(http.HomeAssistantView):
    """Download diagnostics view."""

    url = "/api/diagnostics/{d_type}/{d_id}"
    extra_urls = ["/api/diagnostics/{d_type}/{d_id}/{sub_type}/{sub_id}"]
    name = "api:diagnostics"

    async def get(
        self,
        request: web.Request,
        d_type: str,
        d_id: str,
        sub_type: str | None = None,
        sub_id: str | None = None,
    ) -> web.Response:
        """Download diagnostics."""
        # Validate d_type and sub_type
        try:
            DiagnosticsType(d_type)
        except ValueError:
            return web.Response(status=HTTPStatus.BAD_REQUEST)

        if sub_type is not None:
            try:
                DiagnosticsSubType(sub_type)
            except ValueError:
                return web.Response(status=HTTPStatus.BAD_REQUEST)

        device_diagnostics = sub_type is not None

        hass: HomeAssistant = request.app["hass"]

        if (config_entry := hass.config_entries.async_get_entry(d_id)) is None:
            return web.Response(status=HTTPStatus.NOT_FOUND)

        diagnostics_data: DiagnosticsData = hass.data[DOMAIN]
        if (info := diagnostics_data.platforms.get(config_entry.domain)) is None:
            return web.Response(status=HTTPStatus.NOT_FOUND)

        filename = f"{config_entry.domain}-{config_entry.entry_id}"

        if not device_diagnostics:
            # Config entry diagnostics
            if info.config_entry_diagnostics is None:
                return web.Response(status=HTTPStatus.NOT_FOUND)
            data = await info.config_entry_diagnostics(hass, config_entry)
            filename = f"{DiagnosticsType.CONFIG_ENTRY}-{filename}"
            return await _async_get_json_file_response(
                hass, data, filename, config_entry.domain, d_id
            )

        # Device diagnostics
        dev_reg = async_get(hass)
        if sub_id is None:
            return web.Response(status=HTTPStatus.BAD_REQUEST)

        if (device := dev_reg.async_get(sub_id)) is None:
            return web.Response(status=HTTPStatus.NOT_FOUND)

        filename += f"-{device.name}-{device.id}"

        if info.device_diagnostics is None:
            return web.Response(status=HTTPStatus.NOT_FOUND)

        data = await info.device_diagnostics(hass, config_entry, device)
        return await _async_get_json_file_response(
            hass, data, filename, config_entry.domain, d_id, sub_id
        )


@callback
def async_has_subscription(
    hass: HomeAssistant, domain: str, config_entry_id: str | None = None
) -> bool:
    """Return True if there is a matching diagnostics subscription."""

    diagnostics_data: DiagnosticsData = hass.data[DOMAIN]
    if (
        domain in diagnostics_data.domain_subscriptions
        and diagnostics_data.domain_subscriptions[domain]
    ):
        return True

    if not config_entry_id:
        return False

    if (
        config_entry_id in diagnostics_data.config_entry_subscriptions
        and diagnostics_data.config_entry_subscriptions[config_entry_id]
    ):
        return True

    return False


@callback
def async_log_object(
    hass: HomeAssistant,
    data: Any,
    domain: str,
    config_entry_id: str | None = None,
) -> None:
    """Send diagnostic data to subscribers."""
    diagnostics_data: DiagnosticsData = hass.data[DOMAIN]

    json_data = json.dumps({"data": data}, cls=ExtendedJSONEncoder)

    domain_subs: set[tuple[websocket_api.ActiveConnection, int]] = set()

    if domain in diagnostics_data.domain_subscriptions:
        for conn, msg_id in diagnostics_data.domain_subscriptions[domain]:
            conn.send_message(websocket_api.event_message(msg_id, json_data))
            domain_subs.add((conn, msg_id))

    if not config_entry_id:
        return

    if (
        config_entry_id in diagnostics_data.config_entry_subscriptions
        and diagnostics_data.config_entry_subscriptions[config_entry_id]
    ):
        for conn, msg_id in diagnostics_data.config_entry_subscriptions[
            config_entry_id
        ]:
            if (conn, msg_id) in domain_subs:
                continue
            conn.send_message(websocket_api.event_message(msg_id, json_data))
