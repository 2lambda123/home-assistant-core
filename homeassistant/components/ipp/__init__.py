"""The Internet Printing Protocol (IPP) integration."""
from __future__ import annotations

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import (
    CONF_HOST,
    CONF_PORT,
    CONF_SSL,
    CONF_VERIFY_SSL,
    Platform,
)
from homeassistant.core import HomeAssistant
from homeassistant.helpers.device_registry import async_get as dr_async_get

from .const import CONF_BASE_PATH, DOMAIN
from .coordinator import IPPDataUpdateCoordinator

PLATFORMS = [Platform.SENSOR]


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up IPP from a config entry."""
    # config flow sets this to either UUID, serial number or None
    if (device_id := entry.unique_id) is None:
        device_id = entry.entry_id

    coordinator = IPPDataUpdateCoordinator(
        hass,
        entry,
        host=entry.data[CONF_HOST],
        port=entry.data[CONF_PORT],
        base_path=entry.data[CONF_BASE_PATH],
        tls=entry.data[CONF_SSL],
        verify_ssl=entry.data[CONF_VERIFY_SSL],
        device_id=device_id,
    )

    dev_reg = dr_async_get(hass)
    device_entry = dev_reg.async_get_device(
        identifiers={(DOMAIN, device_id)},
    )

    if device_entry and entry.entry_id not in device_entry.config_entries:
        device_entry = None

    if not device_entry:
        await coordinator.async_config_entry_first_refresh()
    else:
        await coordinator.async_refresh()

    if coordinator.last_update_success:
        coordinator.initialized = True

    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
        hass.data[DOMAIN].pop(entry.entry_id)
    return unload_ok
