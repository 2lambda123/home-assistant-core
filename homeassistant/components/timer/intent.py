"""Intents for the timer integration."""
from __future__ import annotations

import logging

import voluptuous as vol

from homeassistant.const import CONF_ENTITY_ID, CONF_ID
from homeassistant.core import Event, HomeAssistant
from homeassistant.helpers import intent
from homeassistant.helpers.entity_component import EntityComponent

from . import (
    CONF_DURATION,
    CONF_RESTORE,
    DOMAIN,
    EVENT_TIMER_FINISHED,
    SERVICE_START,
    Timer,
)

INTENT_START_DEVICE_TIMER = "HassStartDeviceTimer"

_LOGGER = logging.getLogger()


async def async_setup_intents(hass: HomeAssistant) -> None:
    """Set up the timer intents."""
    intent.async_register(hass, StartDeviceTimerIntent())


async def async_device_timer_finished(hass: HomeAssistant, device_id: str) -> None:
    """Responds to a device when its timer is finished."""


class StartDeviceTimerIntent(intent.IntentHandler):
    """Handle StartDeviceTimer intents."""

    intent_type = INTENT_START_DEVICE_TIMER
    slot_schema = {
        "device_id": str,
        vol.Optional("hours"): int,
        vol.Optional("minutes"): int,
        vol.Optional("seconds"): int,
    }

    async def async_handle(self, intent_obj: intent.Intent):
        """Handle the intent."""
        hass = intent_obj.hass
        slots = self.async_validate_slots(intent_obj.slots)
        device_id = slots["device_id"]["value"]
        entity_id = f"timer.timer_device_{device_id}"

        if "hours" in slots:
            hours = int(slots["hours"]["value"])
        else:
            hours = 0

        if "minutes" in slots:
            minutes = int(slots["minutes"]["value"])
        else:
            minutes = 0

        if "seconds" in slots:
            seconds = int(slots["seconds"]["value"])
        else:
            seconds = 0

        duration = f"{hours:02}:{minutes:02}:{seconds:02}"

        timer_state = hass.states.get(entity_id)
        if timer_state is None:
            # Create timer entity for device
            component: EntityComponent[Timer] = hass.data[DOMAIN]
            timer = Timer(
                {
                    CONF_ID: f"device_{device_id}",
                    CONF_DURATION: duration,
                    CONF_RESTORE: False,
                }
            )
            timer.editable = False

            await component.async_add_entities([timer])
            _LOGGER.debug("Created timer entity: %s", timer.entity_id)

        async def handle_finished(event: Event) -> None:
            await async_device_timer_finished(hass, device_id)

        hass.bus.async_listen_once(EVENT_TIMER_FINISHED, handle_finished)

        _LOGGER.debug(
            "Starting timer with duration=%s, id=%s", duration, timer.entity_id
        )

        await hass.services.async_call(
            DOMAIN,
            SERVICE_START,
            {
                CONF_ENTITY_ID: entity_id,
                CONF_DURATION: duration,
            },
            blocking=True,
        )

        timer_state = hass.states.get(timer.entity_id)
        assert timer_state is not None

        response = intent_obj.create_response()
        response.response_type = intent.IntentResponseType.ACTION_DONE
        response.async_set_states(matched_states=[timer_state])
        return response
