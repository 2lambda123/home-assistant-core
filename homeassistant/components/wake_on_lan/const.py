"""Constants for the Wake-On-LAN component."""

from homeassistant.const import Platform

DOMAIN = "wake_on_lan"

PLATFORMS = [Platform.SWITCH]

CONF_INDEX = "index"
CONF_SWITCH = "switch"
