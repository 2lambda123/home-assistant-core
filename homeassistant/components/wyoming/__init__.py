"""The Wyoming integration."""
from __future__ import annotations

import logging

from homeassistant.components.conversation import async_set_agent, async_unset_agent
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ConfigEntryNotReady

from .const import ATTR_SPEAKER, DOMAIN
from .conversation import WyomingConversationAgent
from .data import WyomingService

_LOGGER = logging.getLogger(__name__)

__all__ = [
    "ATTR_SPEAKER",
    "DOMAIN",
]


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Load Wyoming."""
    service = await WyomingService.create(entry.data["host"], entry.data["port"])

    if service is None:
        raise ConfigEntryNotReady("Unable to connect")

    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = service

    intent_installed = any(intent.installed for intent in service.info.intent)
    handle_installed = any(handle.installed for handle in service.info.handle)

    if intent_installed or handle_installed:
        async_set_agent(hass, entry, WyomingConversationAgent(hass, entry, service))

    await hass.config_entries.async_forward_entry_setups(
        entry,
        service.platforms,
    )

    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload Wyoming."""
    service: WyomingService = hass.data[DOMAIN][entry.entry_id]

    async_unset_agent(hass, entry)

    unload_ok = await hass.config_entries.async_unload_platforms(
        entry,
        service.platforms,
    )
    if unload_ok:
        del hass.data[DOMAIN][entry.entry_id]

    return unload_ok
