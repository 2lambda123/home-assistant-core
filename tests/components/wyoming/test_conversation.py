"""Test tts."""
from __future__ import annotations

from unittest.mock import patch

import pytest
from syrupy.assertion import SnapshotAssertion
from wyoming.handle import Handled, NotHandled
from wyoming.intent import Entity, Intent, NotRecognized

from homeassistant.components import conversation
from homeassistant.const import ATTR_FRIENDLY_NAME
from homeassistant.core import Context, HomeAssistant
from homeassistant.helpers import entity_registry as er, intent
from homeassistant.setup import async_setup_component

from . import MockAsyncTcpClient

from tests.common import async_mock_service


@pytest.fixture
async def init_components(hass: HomeAssistant):
    """Initialize relevant components with empty configs."""
    assert await async_setup_component(hass, "homeassistant", {})
    assert await async_setup_component(hass, "conversation", {})
    assert await async_setup_component(hass, "intent", {})


async def test_intent_recognized(
    hass: HomeAssistant,
    entity_registry: er.EntityRegistry,
    init_components,
    init_wyoming_intent,
    intent_config_entry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test intent recogntion."""
    test_light = entity_registry.async_get_or_create("light", "demo", "1234")
    hass.states.async_set(
        test_light.entity_id, "off", attributes={ATTR_FRIENDLY_NAME: "test light"}
    )
    calls = async_mock_service(hass, "light", "turn_on")

    intent_events = [
        Intent(
            name="HassTurnOn",
            entities=[Entity(name="name", value="test light")],
            text="Turned on light",
        ).event()
    ]

    with patch(
        "homeassistant.components.wyoming.conversation.AsyncTcpClient",
        MockAsyncTcpClient(intent_events),
    ) as mock_client:
        result = await conversation.async_converse(
            hass,
            "Turn on test light",
            conversation_id=None,
            context=Context(),
            agent_id=intent_config_entry.entry_id,
        )
        assert len(calls) == 1
        assert result.response.response_type == intent.IntentResponseType.ACTION_DONE
        assert result.response.speech == {
            "plain": {"speech": "Turned on light", "extra_data": None}
        }

    assert mock_client.written == snapshot


async def test_intent_not_recognized(
    hass: HomeAssistant,
    entity_registry: er.EntityRegistry,
    init_components,
    init_wyoming_intent,
    intent_config_entry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test intent recogntion."""
    intent_events = [NotRecognized("I don't understand").event()]

    with patch(
        "homeassistant.components.wyoming.conversation.AsyncTcpClient",
        MockAsyncTcpClient(intent_events),
    ) as mock_client:
        result = await conversation.async_converse(
            hass,
            "You won't understand this",
            conversation_id=None,
            context=Context(),
            agent_id=intent_config_entry.entry_id,
        )
        assert result.response.response_type == intent.IntentResponseType.ERROR
        assert (
            result.response.error_code == intent.IntentResponseErrorCode.NO_INTENT_MATCH
        )

    assert mock_client.written == snapshot


async def test_intent_recognition_failed(
    hass: HomeAssistant,
    entity_registry: er.EntityRegistry,
    init_components,
    init_wyoming_intent,
    intent_config_entry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test intent recogntion."""
    intent_events = [Intent(name="NotAnIntent").event()]

    with patch(
        "homeassistant.components.wyoming.conversation.AsyncTcpClient",
        MockAsyncTcpClient(intent_events),
    ) as mock_client, pytest.raises(intent.UnknownIntent):
        await conversation.async_converse(
            hass,
            "Do something unexpected",
            conversation_id=None,
            context=Context(),
            agent_id=intent_config_entry.entry_id,
        )

    assert mock_client.written == snapshot


async def test_intent_recognition_client_dropped(
    hass: HomeAssistant,
    entity_registry: er.EntityRegistry,
    init_components,
    init_wyoming_intent,
    intent_config_entry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test intent recogntion."""
    intent_events = [None]

    with patch(
        "homeassistant.components.wyoming.conversation.AsyncTcpClient",
        MockAsyncTcpClient(intent_events),
    ) as mock_client:
        result = await conversation.async_converse(
            hass,
            "This will never be recognized",
            conversation_id=None,
            context=Context(),
            agent_id=intent_config_entry.entry_id,
        )

        assert result.response.response_type == intent.IntentResponseType.ERROR
        assert result.response.error_code == intent.IntentResponseErrorCode.UNKNOWN

    assert mock_client.written == snapshot


async def test_intent_handled(
    hass: HomeAssistant,
    entity_registry: er.EntityRegistry,
    init_components,
    init_wyoming_handle,
    handle_config_entry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test intent recogntion."""
    handle_events = [Handled("Turned on light").event()]

    with patch(
        "homeassistant.components.wyoming.conversation.AsyncTcpClient",
        MockAsyncTcpClient(handle_events),
    ) as mock_client:
        result = await conversation.async_converse(
            hass,
            "Turn on test light",
            conversation_id=None,
            context=Context(),
            agent_id=handle_config_entry.entry_id,
        )
        assert result.response.response_type == intent.IntentResponseType.ACTION_DONE
        assert result.response.speech == {
            "plain": {"speech": "Turned on light", "extra_data": None}
        }

    assert mock_client.written == snapshot


async def test_intent_not_handled(
    hass: HomeAssistant,
    entity_registry: er.EntityRegistry,
    init_components,
    init_wyoming_handle,
    handle_config_entry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test intent recogntion."""
    handle_events = [NotHandled("I don't understand").event()]

    with patch(
        "homeassistant.components.wyoming.conversation.AsyncTcpClient",
        MockAsyncTcpClient(handle_events),
    ) as mock_client:
        result = await conversation.async_converse(
            hass,
            "You won't understand this",
            conversation_id=None,
            context=Context(),
            agent_id=handle_config_entry.entry_id,
        )
        assert result.response.response_type == intent.IntentResponseType.ERROR
        assert (
            result.response.error_code
            == intent.IntentResponseErrorCode.FAILED_TO_HANDLE
        )
        assert result.response.speech == {
            "plain": {"speech": "I don't understand", "extra_data": None}
        }

    assert mock_client.written == snapshot


async def test_intent_handling_client_dropped(
    hass: HomeAssistant,
    entity_registry: er.EntityRegistry,
    init_components,
    init_wyoming_handle,
    handle_config_entry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test intent recogntion."""
    handle_events = [None]

    with patch(
        "homeassistant.components.wyoming.conversation.AsyncTcpClient",
        MockAsyncTcpClient(handle_events),
    ) as mock_client:
        result = await conversation.async_converse(
            hass,
            "This will never be handled",
            conversation_id=None,
            context=Context(),
            agent_id=handle_config_entry.entry_id,
        )
        assert result.response.response_type == intent.IntentResponseType.ERROR
        assert result.response.error_code == intent.IntentResponseErrorCode.UNKNOWN

    assert mock_client.written == snapshot
