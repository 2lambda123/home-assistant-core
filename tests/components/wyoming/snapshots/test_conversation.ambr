# serializer version: 1
# name: test_intent_handled
  list([
    dict({
      'data': dict({
        'text': 'Turn on test light',
      }),
      'payload': None,
      'type': 'transcript',
    }),
  ])
# ---
# name: test_intent_handling_client_dropped
  list([
    dict({
      'data': dict({
        'text': 'This will never be handled',
      }),
      'payload': None,
      'type': 'transcript',
    }),
  ])
# ---
# name: test_intent_not_handled
  list([
    dict({
      'data': dict({
        'text': "You won't understand this",
      }),
      'payload': None,
      'type': 'transcript',
    }),
  ])
# ---
# name: test_intent_not_recognized
  list([
    dict({
      'data': dict({
        'text': "You won't understand this",
      }),
      'payload': None,
      'type': 'recognize',
    }),
  ])
# ---
# name: test_intent_recognition_client_dropped
  list([
    dict({
      'data': dict({
        'text': 'This will never be recognized',
      }),
      'payload': None,
      'type': 'recognize',
    }),
  ])
# ---
# name: test_intent_recognition_failed
  list([
    dict({
      'data': dict({
        'text': 'Do something unexpected',
      }),
      'payload': None,
      'type': 'recognize',
    }),
  ])
# ---
# name: test_intent_recognized
  list([
    dict({
      'data': dict({
        'text': 'Turn on test light',
      }),
      'payload': None,
      'type': 'recognize',
    }),
  ])
# ---
